### ISSUE TITLE
Add a Post-Load Notification to Resources

### ISSUE THREAD

**Issue by operation404** [godotengine/godot-proposals#10219]
### Describe the project you are working on

I'm making an top-down action rpg which has a variety of weapons and abilities with unique hitbox shapes and customization. I have many custom resources representing things varying from the possible stages an ability/attack can be in, how a tween will animate an effect or action, custom hitbox shapes, the statistics of items and creatures, etc.

### Describe the problem or limitation you are having in your project

I often find that I have resources with exported properties that are then used to calculate other values. The majority of the time these values only need to computed once, as many don't update during the game and are only changed in the editor. They can also be computationally complex and/or depend on multiple different properties which may have invalid default states that need to be set before the computation can occur.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a notification to the base Resource class which triggers after it has been completely loaded. This notification only triggers once per resource and is _not_ sent for manually instantiated resources, only those that are loaded via calls to ResourceLoader. The notification is guaranteed to occur only after all properties have been set, including properties which themselves are resources. But the order in which sub-resources receive the notification is _not_ guaranteed, as they might be shared and already loaded.

This would allow developers to perform computations that depend on resource properties in a less error-prone and boilerplate-heavy way than needing to implement custom setters for each property or manually checking if derived values were computed after each load call.

This proposal is related to #8758 and #296, though I have also implemented it in a fork.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I have already implemented this on [my own fork of Godot.](https://github.com/operation404/godot/commit/4ac3752b75f7dc0e621e06c5a7b99c356bf0ffa4) I will also describe how I implemented it here as well though.

Resource has the `NOTIFICATION_POSTLOAD` notification defined as well as a private bool `postload_first` representing whether it has received the notification. A virtual method `_post_load` called when `NOTIFICATION_POSTLOAD` is received is also defined, for convenience.

A public method `send_postload` is defined in Resource:
```cpp
void Resource::send_postload() {
    if (postload_first) {
        postload_first = false;
        notification(NOTIFICATION_POSTLOAD);
        GDVIRTUAL_CALL(_post_load);
    }
}
```

`send_postload` is called in 3 different places:

- In `ResourceLoader::_load`, after the resource has been loaded and verified to be non-null, just before it is returned from the function. The call here ensures that each resource loaded via any ResourceFormatLoader receives the notification, including sub-resources which are stored in separate files.
- In `ResourceLoaderText::load` and `ResourceLoaderBinary::load` after each internal resource has been instantiated and had its properties set. The calls here ensure that internal sub-resources which are stored in the main resource's file receive the notification.
  - Calling `send_postload` from within these classes was the best way I found to ensure internal resources would receive `NOTIFICATION_POSTLOAD`. Calling `send_postload` from the main resource into its sub-resources would involve combing through its properties in both the core code and any potential attached script. Since this is basically already done as part of the loading process, I thought it more straightforward to include the calls there instead. As far as I currently know, these are the only two loaders which handle files which can contain internal resources.

Currently, I do not expose `send_postload` as part of the public extension API for Resource, so it isn't available to be called in GDScript or C#. I am not familiar with the process of writing custom ResourceFormatLoaders, but if it would be useful to have access to this method then I'd be completely fine with exposing it. I forwent doing so purely to minimize the amount of changes I was making.

While I didn't do it in my implementation, I could also see an argument being made that the notification only needs to be sent for resources with user-defined scripts attached since those are the only resources which would make use of such a notification. I found it simpler to not check for that case and don't think it does any harm, but it may be more performant to check depending on how costly sending a notification is. I am nowhere near familiar enough with the Godot engine to determine that myself.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

There are workarounds:

1. Whenever a call to load a resource is made, follow it with a manual check to see if the derived values have been computed.
2. Have custom property setters emit the changed signal and connect a method to it to compute derived values.
3. Have custom property setters directly compute the derived values, or call a method that does so.

While these do work, they involve a lot of additional boilerplate and conditional clutter, and introduce more opportunities for potential user-error. It's already extremely common for developers to use `NOTIFICATION_READY` (or more specifically, `_ready`) to handle any one-time setup code that needs to execute after all of the properties and children of a Node have been set. While not completely identical in behavior, `NOTIFICATION_POSTLOAD` provides a similar way of handling one-time setup calls that occur after all properties are set for Resources.

### Is there a reason why this should be core and not an add-on in the asset library?

It requires modifying the core resource and resource loader classes.

**Comment by Mickeon** [comment]
I am a bit confused as to why overriding `_init` is not enough for most of the use cases.

**Comment by dalexeev** [comment]
* Duplicate of #5994.

> I am a bit confused as to why overriding `_init` is not enough for most of the use cases.

Exported variables are assigned after the resource is created, so the constructor is executed first. As a workaround you could use something like:

```gdscript
@export_storage var _last_exported_property:
    set(_value):
        _setup()
```

**Comment by akien-mga** [comment]
There's a PR open which addresses this proposal: https://github.com/godotengine/godot/pull/109752

It would be helpful to have interested users test it to confirm that it implements what you expect.
